### ISSUE TITLE
Disable server.log saving to disk

### ISSUE THREAD

**Issue by mal0ne-23** [pmmp/PocketMine-MP#1275]
Some people don't use the server.log, like myself. I understand some use it to look for errors on the server, but I have other ways of doing so and the server.log is just an extra file that I don't want there. I don't ever use it and I have a feeling that lots of other people don't use it either.

**Comment by SOF3** [comment]
Writing the server.log uses **harddisk memory** not RAM (random access memory).
PocketMine writes the server.log asynchronously, so it has almost no impact on server performance.
If you are concerned about the server.log taking memory, you may also make a cronjob to clean the server.log periodically.

**Comment by mal0ne-23** [comment]
Ahhhh. But wouldn't it be more convenient to remove it from saving altogether?

**Comment by xxFlare** [comment]
> Writing the server.log to the disk can use up a lot of RAM

You're deluding yourself.

You're also deluding yourself if you think your going to improve performance by disabling it.

There is no benefit to disabling server.log, you won't get any performance improvements or better CPU usage whatsoever, and as SOFE said it is done _asynchronously_ meaning main thread time is not wasted writing to it (file I/O).

Besides from saving a few MB of hard drive space, and most servers have plenty of SSD storage, why would you want to disable something that is really useful for finding console errors during production? For comparison, crash dumps can add up and take a few MB too, but no one in the right mind would disable that.

**Comment by mal0ne-23** [comment]
> Ahhhh.

Kinda indicated that I didn't know it was saved asynchronously. Sorry I didn't know that, now I do and you don't need to repeat to me what I literally JUST read above.

And, maybe I don't want logs of the server saving taking up disk space. I haven't looked at my server logs for my current server, ever. If I suspect a bug I just look through the server's screen session, therefore, O do not **NEED** the server logs. Yeah I understand **YOU** might use it to find console errors. Cool, guess what? I have a preferred method of doing things and don't do the same thing as you, surprising right?

Don't say "oh you're not in the right mind because you want to disable logs", I'm completely in the right mind, I just do things differently I'm sorry. The benefit to me is, there aren't useless server.log files with bullshit that I am never going to look at. Did that make more sense to you? K.


Thank you for the suggestion on how to remove it currently myself, SOFE. But I'd like some way in the pocketmine.yml or server.properties to disable logs from saving. I know I could do cronjobs, but I feel like other people would like a way of disabling the server.log saving aswell. Not saying its necessarily the goto if you aren't very well with the PocketMine API and _constantly_ need to go and check the server.log for errors with plugins. But some people don't need it.

**Comment by Muqsit** [comment]
@SOF3 But what if the server is being hosted on a single core, this brings us back to the "don't abuse asynchronous tasks" topic.

**Comment by Muqsit** [comment]
@SOF3 But what if the server is being hosted on a single core. This brings us back to the "don't abuse asynchronous tasks because they delay pocketmine's chunk sending", just without the abuse.

**Comment by madecode15** [comment]
#1164 - we been here before

@Muqsit You should read dktapps comment there.

@mal0ne-23 hostile much?

**Comment by QuanFromChina** [comment]
@mal0ne-23 server.log lag server ! Good find ! I think that is why CPU always high !

we must add disable option ... not everyone need extra lag for production ...

**Comment by dktapps** [comment]
Duplicate of #1164 

@Muqsit being single-core has nothing to do with anything - disk I/O _doesn't use the CPU_.

This would be particularly inconvenient to implement properly as it would require loading configuration before the Server is created. This is problematic because of the way Config depends on Server for various things.

Spoons such as Genisys hacked this by having the logger thread quit partway through startup, leaving it in a stopped, un-joined state. This is an unpleasant solution and as such will not be considered.

TL;DR: this can't be done without hacks.

**Comment by Muqsit** [comment]
My bad, that Steadfast README should've elaborated a bit on this topic.

**Comment by dktapps** [comment]
To clarify, the logger has its own dedicated thread - it doesn't use asynchronous tasks. Therefore there is no impact to server performance because the dedicated thread is supposed to sit and wait anyway.

**Comment by TheDiamondYT1** [comment]
@dktapps When pmmp/PocketMine-MP is a spoon so you can shut your mouth

**Comment by dktapps** [comment]
@TheDiamondYT1 excuse me?

It appears there was a bug in the MainLogger causing disk writes to be synchronized with the calling thread. This actually _did_ waste main thread time. The bug has been fixed as of https://github.com/pmmp/PocketMine-MP/commit/4e9af1ac4561c5fdca8769c7ffe517bbbad8435f.
